### PR TITLE
[FEATURE][3] Add Favorites

### PR DESCRIPTION
**Pull request by mathiramilo-lit**
# Add Favorites - 3

## Description ⭐

I've implemented add to favorites storing favorite products in local storage.
I've also rebase and merge the URL state handling PR.

### Type of change 🏃

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -35,6 +55,21 @@ function App() {
...
+  useEffect(() => {
+    setShowFavorites(false);
+  }, [debouncedQuery]);
+
+  useEffect(() => {
+    const newParams = new URLSearchParams();
+
+    if (query) newParams.set('search', query);
+    if (sort.sortBy) newParams.set('sortBy', sort.sortBy);
+    if (sort.order) newParams.set('order', sort.order);
```
Comment (danielaWaldeck95) [inline src/App.tsx]:
```
    query && newParams.set('search', query);
    sort.sortBy && newParams.set('sortBy', sort.sortBy);
    sort.order && newParams.set('order', sort.order);
```

File: src/components/ui/loader.tsx [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+import { cn } from '../../lib/utils';
+
+type LoaderProps = ComponentPropsWithoutRef<'div'>;
+
+export const Loader = ({ className }: LoaderProps) => {
+  return (
+    <div
+      className={cn(
+        'h-4 w-4 animate-spin rounded-full border-t-2 border-t-white',
```
Comment (danielaWaldeck95) [inline src/components/ui/loader.tsx]: https://tailwindcss.com/docs/size you can replace h-4 w-4 by size-4

File: src/components/ui/sort-by.tsx [hunk 3]
```diff
@@ -1,17 +1,39 @@
...
+          (opt) => opt.sortBy === sort.sortBy && opt.order === sort.order,
+        )?.label
+      : 'Sort By',
+  );
+
+  useEffect(() => {
+    if (!showFavorites && !sort.sortBy && !sort.order) {
+      setLabel('Sort By');
+    }
+  }, [showFavorites, sort]);
```
Comment (danielaWaldeck95) [inline src/components/ui/sort-by.tsx]: I think this useEffect is not needed, the component will re-render if showFavorites and sort states change on the parent

File: src/hooks/use-products.ts [hunk 4]
```diff
@@ -29,7 +30,7 @@ export const useProducts = ({ query, sort }: UseProductsOptions) => {
 
     try {
       const res = await axios.get(
-        `https://dummyjson.com/products${query ? `/search?q=${query}&` : '?'}limit=${LIMIT}&skip=${pagination.skip}${sort ? `&sortBy=${sort.sortBy}&order=${sort.order}` : ''}`,
+        `https://dummyjson.com/products${query ? `/search?q=${query}&` : '?'}limit=${LIMIT}&skip=${pagination.skip}${sort?.sortBy && sort.order ? `&sortBy=${sort.sortBy}&order=${sort.order}` : ''}`,
```
Comment (danielaWaldeck95) [inline src/hooks/use-products.ts]: same suggestion H left on the other pr

File: src/hooks/use-products.ts [hunk 5]
```diff
@@ -50,9 +51,8 @@ export const useProducts = ({ query, sort }: UseProductsOptions) => {
 
       try {
         const res = await axios.get(
-          `https://dummyjson.com/products${query ? `/search?q=${query}&` : '?'}limit=${LIMIT}${sort ? `&sortBy=${sort.sortBy}&order=${sort.order}` : ''}`,
+          `https://dummyjson.com/products${query ? `/search?q=${query}&` : '?'}limit=${LIMIT}${sort?.sortBy && sort.order ? `&sortBy=${sort.sortBy}&order=${sort.order}` : ''}`,
```
Comment (danielaWaldeck95) [inline src/hooks/use-products.ts]: same suggestion H left on the other pr

File: src/hooks/use-debounce.ts [hunk 6]
```diff
@@ -1,17 +1,21 @@
...
+export function useDebounce<T>(
+  initialValue: T,
+  time: number,
+): [T, T, React.Dispatch<T>] {
+  const [value, setValue] = useState<T>(initialValue);
+  const [debouncedValue, setDebouncedValue] = useState<T>(initialValue);
 
   useEffect(() => {
-    const handler = setTimeout(() => {
+    const debounce = setTimeout(() => {
```
Comment (danielaWaldeck95) [inline src/hooks/use-debounce.ts]: rename to timeoutId

**Review by danielaWaldeck95 — COMMENTED** [review]
Left some comments

File: src/components/products/product-card.tsx [hunk 7]
```diff
@@ -1,14 +1,19 @@
-import { Heart } from '../assets';
-import { Product } from '../types';
+import { Product } from '../../types';
+import { HeartFill, HeartOutline } from '../../assets';
+import { useFavorites } from '../../context';
+import { useMemo } from 'react';
```
Comment (jsarrolt) [inline src/components/products/product-card.tsx]: Consider setting up [typescript barrel generator](https://marketplace.visualstudio.com/items?itemName=eliostruyf.vscode-typescript-exportallmodules) and [path aliases](https://blog.logrocket.com/using-path-aliases-cleaner-react-typescript-imports/) for the project

File: src/components/products/product-card.tsx [hunk 8]
```diff
@@ -1,14 +1,19 @@
...
 }
 
 export const ProductCard = ({ product }: ProductCardProps) => {
-  const handleAddToFavorites = () => {
-    alert(`Product ${product.id} added to favorites!`);
-  };
+  const { favorites, addFavorite, removeFavorite } = useFavorites();
+
+  const isFavorite = useMemo(
+    () => Boolean(favorites.find((p) => p.id === product.id)),
```
Comment (jsarrolt) [inline src/components/products/product-card.tsx]:
```suggestion
    () => favorites.some((p) => p.id === product.id),
```

File: src/components/ui/sort-by.tsx [hunk 9]
```diff
@@ -1,17 +1,39 @@
...
+import { AltArrowDown, SortVertical } from '../../assets';
+import { cn } from '../../lib/utils';
+import { Order, SortBy as SortByType, SortOption } from '../../types';
+import { SortState } from '../../app';
 
 interface SortByProps {
   options: SortOption[];
-  setSort: (sortBy: ProductField, order: Order) => void;
+  sort: SortState;
+  setSort: (sortBy: SortByType | undefined, order: Order | undefined) => void;
```
Comment (jsarrolt) [inline src/components/ui/sort-by.tsx]:
```suggestion
setSort: (sortBy?: SortByType, order?: Order) => void;
```

File: src/App.tsx [hunk 10]
```diff
@@ -11,17 +11,37 @@ import {
...
 import { SORT_OPTIONS } from './constants';
 import { useProducts, useDebounce } from './hooks';
-import { SortOption } from './types';
+import { useFavorites } from './context';
+import { Order, SortBy as SortByType, isOrder, isSortBy } from './types';
 
-export type SortState = Omit<SortOption, 'label'>;
+export type SortState = {
+  sortBy: SortByType | undefined;
+  order: Order | undefined;
```
Comment (jsarrolt) [inline src/App.tsx]:
```suggestion
  sortBy?: SortByType;
  order?: Order;
```

**Review by jsarrolt — COMMENTED** [review]
Great work!

File: src/components/products/product-card.tsx (hunk 7 above)
Comment (mathiramilo-lit) [reply src/components/products/product-card.tsx]: Thanks for the suggestion! I went ahead and added the path aliases

File: src/App.tsx (hunk 1 above)
Comment (mathiramilo-lit) [reply src/App.tsx]:
```
Expected an assignment or function call and instead saw an expression
.eslint[@typescript-eslint/no-unused-expressions](https://typescript-eslint.io/rules/no-unused-expressions)
```

I'm encountering the error "Expected an assignment or function call and instead saw an expression" from ESLint. What could be causing this?

File: src/components/ui/loader.tsx (hunk 2 above)
Comment (mathiramilo-lit) [reply src/components/ui/loader.tsx]: Thanks for the tip! I didn’t know about that Tailwind shortcut.

File: src/hooks/use-products.ts (hunk 4 above)
Comment (mathiramilo-lit) [reply src/hooks/use-products.ts]: Great suggestion! I'm fixing that.

File: src/App.tsx (hunk 1 above)
Comment (danielaWaldeck95) [reply src/App.tsx]: Sorry ESLint is right, mybad, the && conditional is being flagged as an unused expression.

Just keep the ifs. I would add the {} anyways, it adds a few more codelines but prevents errors and improves readability

Comment (mathiramilo-lit) [reply src/App.tsx]: Perfect. So I'll keep the if statements and add the {} for improved readability and error prevention. Thanks for the help!

File: src/components/ui/sort-by.tsx (hunk 3 above)
Comment (mathiramilo-lit) [reply src/components/ui/sort-by.tsx]: The purpose of this useEffect is to execute when the showFavorites state changes in the App component. If I remove it, the functionality stops working. I haven’t been able to figure out a cleaner approach. Do you have any suggestions?

**Pull request merged**